util/network-devp2p/src/handshake.rs

```diff
@@ -314,27 +311,26 @@ impl Handshake {
         Message: Send + Clone + Sync + 'static,
     {
         trace!(target: "network", "Sending handshake auth to {:?}", self.connection.remote_addr_str());
-        let mut data = [0u8; /*Signature::SIZE*/ 65 + /*H256::SIZE*/ 32 + /*Public::SIZE*/ 64 + /*H256::SIZE*/ 32 + 1]; //TODO: use associated constants
-        let len = data.len();
-        {
-            data[len - 1] = 0x0;
-            let (sig, rest) = data.split_at_mut(65);
-            let (hepubk, rest) = rest.split_at_mut(32);
-            let (pubk, rest) = rest.split_at_mut(64);
-            let (nonce, _) = rest.split_at_mut(32);
+        let mut rlp = RlpStream::new_list(4);
+        let shared = *ecdh::agree(secret, &self.id)?;
+        rlp.append(&sign(self.ecdhe.secret(), &(shared ^ self.nonce))?.to_vec());
+        rlp.append(public);
+        rlp.append(&self.nonce);
+        rlp.append(&PROTOCOL_VERSION);
 
-            // E(remote-pubk, S(ecdhe-random, ecdh-shared-secret^nonce) || H(ecdhe-random-pubk) || pubk || nonce || 0x0)
-            let shared = *ecdh::agree(secret, &self.id)?;
-            sig.copy_from_slice(&*sign(self.ecdhe.secret(), &(shared ^ self.nonce))?);
-            write_keccak(self.ecdhe.public(), hepubk);
-            pubk.copy_from_slice(public);
-            nonce.copy_from_slice(&self.nonce);
-        }
-        let message = ecies::encrypt(&self.id, &[], &data)?;
-        self.auth_cipher = message.clone();
-        self.connection.send(io, message);
-        self.connection.expect(V4_ACK_PACKET_SIZE);
-        self.state = HandshakeState::ReadingAck;
+        let encoded = rlp
+            .out()
+            .into_iter()
+            .chain(repeat_with(rand::random).take(rand::thread_rng().gen_range::<usize>(100, 301)))
```

Comment:
actually looking at what geth does (although my `go` skills are weak), it seems to be zeroes in range `[100, 300)`
https://github.com/ethereum/go-ethereum/blob/97fc1c3b1d054a1345934e18fa6c3489e9119666/p2p/rlpx/rlpx.go#L592

Reply:
@ordian True, but implementations shouldn't read or touch this padding anyway.

Reply:
It's OK to have random padding, but the padding will be encrypted, so it's also OK (and cheaper) to use zeros instead.

Reply:
Changed to zeroes